script/search/2.0/search.inc.php

```diff
@@ -0,0 +1,357 @@
+      return $this->WriteSearchResults($result);
+    }
+
+    private function GetSearchQueries(KeyboardSearchResult $result) {
+      $result->searchtext = strip_tags($result->searchtext);
+
+      switch($result->context) {
+      case KeyboardSearchResult::CONTEXT_KEYBOARD:
+        return $this->LoadKeyboardSearch($result);
+
+      case KeyboardSearchResult::CONTEXT_LANGUAGE:
+        switch($result->filter) {
+        case KeyboardSearchResult::FILTER_ALL:
+          $result->rangetext = "All languages matching '{$result->searchtext}'";
+          $result->languages = $this->LoadLanguageSearch($result->text, 1, $result->allmatch);
+          break;
+        case KeyboardSearchResult::FILTER_ID:
+          $result->rangetext = "Languages with BCP 47 code '{$result->searchtext}'";
+          $result->languages = $this->LoadLanguageSearch($result->text, 1, $result->allmatch);
+          break;
+        case KeyboardSearchResult::FILTER_SEARCH:
+          $result->rangetext = "Languages matching '{$result->searchtext}'";
+          $result->languages = $this->LoadLanguageSearch($result->text, 0, $result->allmatch); //?
+          break;
+        }
+        break;
+
+      case KeyboardSearchResult::CONTEXT_REGION:
+        if($result->regionmatch) {
+          $result->rangetext = "Countries in {$result->region}";
+          $result->countries = $this->LoadRegionSearch($result->text, $result->regionmatch ? 2 : 1);
+        }
+        //else if($result->allmatch && $result->idmatch) $rangetext = "All languages for country with ISO3166-1 code '{$result->searchtext}'";
+        //else if($result->allmatch) $rangetext = "Countries matching '{$result->searchtext}'";
+        else if($result->idmatch) {
+//          $rangetext = "Languages for country with ISO3166-1 code '{$result->searchtext}'";
+//          $langs = $this->LoadLanguageSearch($result->text, 2, $result->allmatch);
+          $result->countries = $this->LoadRegionSearch($result->text, $result->regionmatch ? 2 : 1);
+        } else {
+          $result->rangetext = "Countries matching '{$result->searchtext}'";
+          $result->countries = $this->LoadRegionSearch($result->text, $result->regionmatch ? 2 : 1);
+        }
+        break;
+      }
+
+      return true;
+    }
+
+    private function WriteSearchResults(KeyboardSearchResult $result) {
+      $this->GetSearchQueries($result);
+
+      $data = array();
+
+      $totalPages = round(($result->totalRows + $result->pageSize - 1)/$result->pageSize);
+
+      // TODO: fixup schema
+
+      $data['context'] = [
+        'range' => $result->rangetext,
+        'context' => $result->context,
+        'pageSize' => $result->pageSize,
+        'pageNumber' => $result->pageNumber,
+        'totalRows' => $result->totalRows,
+        'totalPages' => $totalPages
+      ];
+
+      if($result->platform !== null)
+        $data['context']['platform'] = $result->platform;
+
+      // TODO: include weighting, matched term, popularity values
```

Comment:
Is this still a TODO?  The PR's title makes me think otherwise... but then again, I don't see any code here that would directly do this.

So, where does the weighting come into play?

Reply:
Yeah, this is now implemented. You can see examples in the fixtures. Each result has a `match` object, e.g.

```
 "match": {
                "name": "Khmer",
                "type": "language",
                "weight": "160",
                "downloads": "37",
                "final_weight": "742.01378555622171"
            }
```

This describes the matched phrase, the type of match (language, keyboard, etc), the raw weight of the match and the composite weight (see `sp_keyboard_search.sql` for how weight is calculated; composite weight accounts for popularity, raw weight is based on search term matching only). Downloads is not 100% correct because only accounts for specific download pathways but will become more accurate over time as we remove the old download patterns.

Reply:
OK... I can see that part now.  I'm having trouble seeing where sorting by weight comes in to play... or is that built into a query somewhere?  Did I miss it in an `orderBy` section somewhere?

Where/how is this weight used?

Reply:
The weighting is used only in the stored procedure but reported back to the client in case it is useful. The results as they come out are in the right order already. keyman.com will use some of the other fields but doesn't currently use the weight fields.